Option to ignore object validity check/refetching if cache is available

shahor02 opened the pull request:
When used with the DPL CCDB fetcher (i.e. loadFileToMemory is called) in order to prefer the available
snapshot w/o its validity check an extra variable `IGNORE_VALIDITYCHECK_OF_CCDB_LOCALCACHE` must be defined
otherwhise the object will be fetched from the server after the validity check and new snapshot will be created if needed.
`IGNORE_VALIDITYCHECK_OF_CCDB_LOCALCACHE` should be always accompanied by the `ALICEO2_CCDB_LOCALCACHE` env.var indicating the snapshot directory (if the directory of particular snapshots are absent, they will be populated at 1st call).

@sawenzel

shahor02 committed:
Option to ignore object validity check/refetching if cache is available

When used with the DPL CCDB fetcher (i.e. loadFileToMemory is called) in order to prefer the available
snapshot w/o its validity check an extra variable IGNORE_VALIDITYCHECK_OF_CCDB_LOCALCACHE must be defined
otherwhise the object will be fetched from the server after the validity check and new snapshot will be created if needed.
IGNORE_VALIDITYCHECK_OF_CCDB_LOCALCACHE should be always accompanied by the ALICEO2_CCDB_LOCALCACHE env.var
indicating the snapshot directory (if the directory of particular snapshots are absent, they will be populated at 1st
call).

sawenzel commented:
It appears to be working. Although, the logging to `ALICEO2_CCDB_LOCALCACHE/log` seems no longer happening.
Before, I could see who is reading from the cache.

shahor02 commented:
Do you want to create a semaphore just to update the log file with the info on reading? Why not simply producing a `LOG(info)`, probably conditioned by the env.var?
BTW, the semaphore you use associated with the object path being updated and the log updater is in the same scope. This will not protect against the simultaneous log file update if two concurrent processes create semaphores for 2 different objects, right?

sawenzel commented:
I am just asking about features/behaviour which were there before and on which I was relying. The technical implementation is a different matter.

sawenzel commented:
So I take it you no longer put a semaphore at all? So what happens if 2 processes download the same object at the same time?

shahor02 commented:
The semaphore is there for storing the object, reading does not need it.
As I wrote, I am not against the logging, but why do you need it to go to separate file which would cost a dedicated semaphore?

ktf commented:
why using a semaphore? moving files is guaranteed to be atomic. can't you just download the file somewhere unique and move it in place?

shahor02 committed:
Log all files loaded from the server of snapshot.

All files read from the real server or local snapshot are reported in the log
with the loading mode (retrieve object or load blob to memory) and CcdbApi
unique instance ID.

shahor02 commented:
@sawenzel shall I merge this?

sawenzel commented:
Ok for me.
